sdk/python/feast/infra/offline_stores/contrib/spark_offline_store/spark.py

```diff
@@ -0,0 +1,522 @@
+    SparkSource,
+)
+from feast.infra.offline_stores.offline_store import (
+    OfflineStore,
+    RetrievalJob,
+    RetrievalMetadata,
+)
+from feast.registry import Registry
+from feast.repo_config import FeastConfigBaseModel, RepoConfig
+from feast.saved_dataset import SavedDatasetStorage
+from feast.type_map import spark_schema_to_np_dtypes
+
+
+class SparkOfflineStoreConfig(FeastConfigBaseModel):
+    type: StrictStr = "spark"
+    """ Offline store type selector"""
+
+    spark_conf: Optional[Dict[str, str]] = None
+    """ Configuration overlay for the spark session """
+    # sparksession is not serializable and we dont want to pass it around as an argument
+
+
+class SparkOfflineStore(OfflineStore):
+    @staticmethod
+    def pull_latest_from_table_or_query(
+        config: RepoConfig,
+        data_source: DataSource,
+        join_key_columns: List[str],
+        feature_name_columns: List[str],
+        event_timestamp_column: str,
+        created_timestamp_column: Optional[str],
+        start_date: datetime,
+        end_date: datetime,
+    ) -> RetrievalJob:
+        spark_session = get_spark_session_or_start_new_with_repoconfig(
+            config.offline_store
+        )
+        assert isinstance(config.offline_store, SparkOfflineStoreConfig)
+        assert isinstance(data_source, SparkSource)
+
+        warnings.warn(
+            "The spark offline store is an experimental feature in alpha development. "
+            "Some functionality may still be unstable so functionality can change in the future.",
+            RuntimeWarning,
+        )
+
+        print("Pulling latest features from spark offline store")
+
+        from_expression = data_source.get_table_query_string()
+
+        partition_by_join_key_string = ", ".join(join_key_columns)
+        if partition_by_join_key_string != "":
+            partition_by_join_key_string = (
+                "PARTITION BY " + partition_by_join_key_string
+            )
+        timestamps = [event_timestamp_column]
+        if created_timestamp_column:
+            timestamps.append(created_timestamp_column)
+        timestamp_desc_string = " DESC, ".join(timestamps) + " DESC"
+        field_string = ", ".join(join_key_columns + feature_name_columns + timestamps)
+
+        start_date_str = _format_datetime(start_date)
+        end_date_str = _format_datetime(end_date)
+        query = f"""
+                SELECT
+                    {field_string}
+                    {f", {repr(DUMMY_ENTITY_VAL)} AS {DUMMY_ENTITY_ID}" if not join_key_columns else ""}
+                FROM (
+                    SELECT {field_string},
+                    ROW_NUMBER() OVER({partition_by_join_key_string} ORDER BY {timestamp_desc_string}) AS feast_row_
+                    FROM {from_expression} t1
+                    WHERE {event_timestamp_column} BETWEEN TIMESTAMP('{start_date_str}') AND TIMESTAMP('{end_date_str}')
```

Comment:
I'm not sure if it was intentional that both the start and the end are inclusive, but IMO the starting point should be inclusive and the ending point should be exclusive. That way, if you have ranges of tuples like (a, b) and (b, c), they are contiguous and non-overlapping. By using `BETWEEN` the ranges will instead overlap at `b`.